### PR TITLE
do JSON comparison of old value/new value

### LINKED ISSUE #3475: Editing a node with select-multiple causes the node changed status to be set to true even when no values of the node have changed
### Current Behavior

Editing a node with select-multiple causes the node `changed` status to be set to `true` even when no values of the node have changed.

https://github.com/node-red/node-red/blob/fdb868516f011044b7369cf3bacfb88fc018aa32/packages/node_modules/%40node-red/editor-client/src/js/ui/editors/panes/properties.js#L110

The above condition doesn't take into account when the values are not primitives.

A fix working for me, based how on the [handleEditSave](https://github.com/node-red/node-red/blob/fdb868516f011044b7369cf3bacfb88fc018aa32/packages/node_modules/%40node-red/editor-client/src/js/ui/editor.js#L647) handles it
```js
if (!isEqual(node[d], newValue)) {

...


function isEqual(oldValue, newValue) {
    if (typeof newValue === "string" || typeof newValue === "number") {
        return oldValue == newValue;
    }
    if (typeof newValue === "boolean") {
        return oldValue === newValue;
    }

    return JSON.stringify(oldValue) === JSON.stringify(newValue);
}
```

side note:

Pretty sure this will never be true as nodeName is always uppercase but I also believe it is not needed in the current version as an empty select returns an empty array not null.

https://github.com/node-red/node-red/blob/fdb868516f011044b7369cf3bacfb88fc018aa32/packages/node_modules/%40node-red/editor-client/src/js/ui/editors/panes/properties.js#L53-L59

### Expected Behavior

Be able to open a node that contains a select element in the editor and close it without the change flag being set to true.

### Steps To Reproduce

_No response_

### Example flow

```
paste your flow here
```

### Environment

- Node-RED version: 2.2.2
- Node.js version: 14.19.0
- npm version: 6.14.16
- Platform/OS: Linux 4.19.104-microsoft-standard x64 LE
- Browser:

### PR DESCRIPTION
**Pull request by Steve-Mcl**
fixes #3475

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Proposed changes

Use `JSON.stringify` to compare old value to new value. This takes care of array and object values & prevents the node signalling a change (thus preventing the deploy button becoming enabled)

More details in issue #3475 

## Checklist
- [x] I have read the [contribution guidelines](https://github.com/node-red/node-red/blob/master/CONTRIBUTING.md)
- [ ] For non-bugfix PRs, I have discussed this change on the forum/slack team.
- [x] I have run `grunt` to verify the unit tests pass
- [ ] I have added suitable unit tests to cover the new/changed functionality

### CODE REVIEW

File: packages/node_modules/@node-red/editor-client/src/js/ui/editors/panes/properties.js [hunk 1]
```diff
@@ -107,7 +107,7 @@
                                         newValue = "";
                                     }
                                 }
-                                if (node[d] != newValue) {
+                                if (!JSON.stringify(node[d]) ===  JSON.stringify(newValue)) {
```
Comment (knolleary) [inline packages/node_modules/@node-red/editor-client/src/js/ui/editors/panes/properties.js:110]:
```suggestion
                                if (JSON.stringify(node[d]) !==  JSON.stringify(newValue)) {
```

**Comment by Steve-Mcl** [comment]
@knolleary I have decided we should (at this point in the flow) handle any errors raised from undefined or BigInt values. 

What I haven't done is `console.log` the error in the internal `isEqual` helper function. 

If you think it makes more sense to  go with the inline JSON compare & let errors bubble up I'll change it back.

**Review by knolleary — APPROVED** [review]

**Comment by knolleary** [comment]
Looks good. Any circular ref is going to break things when the node is exported as part of deploy - so should not be a 'normal' thing to handle at all.

**Comment by zachowj** [comment]
When a node has a default value that is a number this solution will cause the same problem it's trying to fix. 

When a value is read from `.val()` it will be a string but the old value was a number before being placed into the text field. This will fail both the strict equality and JSON comparison.

https://github.com/node-red/node-red/blob/ac97e8c6136da712a53f6c024a3f7d78d6ba3ec7/packages/node_modules/%40node-red/nodes/core/common/20-inject.html#L247

https://jsfiddle.net/b2Lox05z/1/

**Comment by knolleary** [comment]
@zachowj well spotted. Yes - we still need to handle the 'almost' equal case between string representations of numbers

**Comment by Steve-Mcl** [comment]
@zachowj please take a look at 10f77fdf1a9d9f87576f901ab523040078008395 - I have added non strict comparison for number / string & let me know if this satisfies the requirement.

**Comment by zachowj** [comment]
Yes, it works for what I have tested.

One edge case it doesn't work for would be `false == ""` but with my limited knowledge of NR, I am not sure if that needs to be checked for here.

**Comment by Steve-Mcl** [comment]
@knolleary before we merge this, I have one last consideration...

The original code was a non strict check & apart from the OPs issue, worked for for a long time.

So, I suspect we should simply stick with non strict comparison + the JSON.stringify as the OP suggests to handle objects, arrays, dates?

```js
function isEqual(originalValue, newValue) {
        try {
            if(originalValue == newValue) {
                return true; 
            }
            return JSON.stringify(originalValue) === JSON.stringify(newValue);
        } catch (err) {
            return false;
        }
}
```

Then, if/when necessary, a future iteration could then handle BigInt and anything else that trips it up.

**Comment by Steve-Mcl** [comment]
@zachowj @knolleary I will assume the final proposal is ok and push the change to close this one off

> The original code was a non strict check & apart from the OPs issue, worked for for a long time.
> So, I suspect we should simply stick with non strict comparison + the JSON.stringify as the OP suggests to handle objects, arrays, dates?
> 
> ```js
> function isEqual(originalValue, newValue) {
>         try {
>             if(originalValue == newValue) {
>                 return true; 
>             }
>             return JSON.stringify(originalValue) === JSON.stringify(newValue);
>         } catch (err) {
>             return false;
>         }
> }
> ```

Commit incoming :)

**Comment by knolleary** [comment]
Yup - looks good.

**Pull request merged**